File: sdk/py/thicket/record.py
"""Fiber records: build, sign, and verify — field order matches the spec."""

from __future__ import annotations

from . import crypto


def _sorted_map(m: dict) -> dict:
    # String→String maps are emitted with keys in ascending order.
    return dict(sorted(m.items()))


def capability(kind, description, *, io=None, tags=None, modalities=None, envelope=None):
    c = {"kind": kind, "description": description}
    if io is not None:
        c["io"] = io
    c["tags"] = list(tags or [])
    c["modalities"] = list(modalities or [])
    c["envelope"] = _sorted_map(envelope or {})
    return c


def locator(protocol, endpoint):
    return {"protocol": protocol, "endpoint": endpoint}


def lease(ttl, issued_at, expires_at):
    return {"ttl": ttl, "issued_at": issued_at, "expires_at": expires_at}


def build_record_payload(
    *,
    schema,
    root: crypto.RootKey,
    endorsement: dict,
    kind,
    locators=None,
    capabilities=None,
    profile=None,
    supports=None,
    visibility="Public",
    lease=None,
    version,
    ext=None,
) -> dict:
    """Build a RecordPayload as an ordered dict (declaration order, lease omitted
    when absent)."""
    payload = {
        "schema": schema,
        "id": root.id(),
        "root_public_key": root.public(),
        "keys": [endorsement],
        "kind": kind,
        "locators": list(locators or []),
        "capabilities": list(capabilities or []),
        "profile": _sorted_map(profile or {}),
        "supports": _sorted_map(supports or {}),
        "visibility": visibility,
    }
    if lease is not None:
        payload["lease"] = lease
    payload["version"] = version
    payload["ext"] = _sorted_map(ext or {})
    return payload


def sign_record(payload: dict, working: crypto.WorkingKey) -> dict:
    sig = working.sign(crypto.signing_input("thicket-record-v1", payload))
    return {"payload": payload, "signer_pub": working.public(), "signature": sig}


def self_record(
    local,
    *,
    kind,
    capabilities,
    locators,
    schema="thicket/record/1",
    visibility="Public",
    lease_ttl=3600,
    version=1,
    profile=None,
    supports=None,
    ext=None,
) -> dict:
    """A fiber builds and signs a record for its own identity."""
    from .identity import unix_now

    now = unix_now()
    payload = build_record_payload(
        schema=schema,
        root=local.root,
        endorsement=local.endorsements[0],
        kind=kind,
        locators=locators,
        capabilities=capabilities,
        profile=profile,
        supports=supports,
        visibility=visibility,
        lease=lease(lease_ttl, now, now + lease_ttl),
        version=version,
        ext=ext,
    )
    return sign_record(payload, local.working)


def verify_record(signed: dict, now: int) -> bool:
    payload = signed["payload"]
    # 1. id binds to the root public key
    if payload["id"] != crypto.sha256(payload["root_public_key"]):
        return False
    # 2. signer is a root-endorsed working key, valid at `now`
    signer = signed["signer_pub"]
    endo = next((e for e in payload["keys"] if e["working_pub"] == signer), None)
    if endo is None:
        return False
    view = {
        "working_pub": endo["working_pub"],
        "not_before": endo["not_before"],
        "not_after": endo["not_after"],
    }
    if not crypto.verify_sig(
        payload["root_public_key"],
        crypto.signing_input("thicket-endorsement-v1", view),
        endo["root_sig"],
    ):
        return False
    if now < endo["not_before"] or now > endo["not_after"]:
        return False
    # 3. signature over the canonical payload
    return crypto.verify_sig(
        signer, crypto.signing_input("thicket-record-v1", payload), signed["signature"]
    )

File: crates/thicket-core/src/record.rs
//! The Resource Record: the signed document a resource publishes (plan §2).
//!
//! A record is split into a [`RecordPayload`] (everything that is signed) and
//! the surrounding [`SignedRecord`] (the signer's working key + the signature).
//! This avoids a struct that signs over its own signature, and makes the
//! canonical signing input unambiguous.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::capability::{Capability, Lease, Locator, Visibility};
use crate::crypto::{signing_bytes, verify_raw};
use crate::error::{Error, Result};
use crate::identity::{verify_working_key, Id, KeyEndorsement, RevocationSet, WorkingKey};

const RECORD_DOMAIN: &str = "thicket-record-v1";

/// The signed body of a resource record.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct RecordPayload {
    /// Record schema + crypto scheme tag (multicodec-style, swappable).
    pub schema: String,
    /// Self-certifying identity = `sha256(root_public_key)`.
    pub id: Id,
    /// The root public key, so a verifier can check `id == sha256(this)`.
    #[serde(with = "serde_bytes")]
    pub root_public_key: Vec<u8>,
    /// Root-signed endorsements of the working keys this resource uses.
    pub keys: Vec<KeyEndorsement>,
    /// Open-set resource kind: model | memory | tool | trigger | agent | …
    pub kind: String,
    #[serde(default)]
    pub locators: Vec<Locator>,
    #[serde(default)]
    pub capabilities: Vec<Capability>,
    /// Perf/economic profile (cost, latency, …). Advisory, string-encoded so it
    /// stays out of the float-canonicalization hazard in the signed form.
    #[serde(default)]
    pub profile: BTreeMap<String, String>,
    /// What the resource speaks, for negotiation (patterns, codecs, …).
    #[serde(default)]
    pub supports: BTreeMap<String, String>,
    #[serde(default)]
    pub visibility: Visibility,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub lease: Option<Lease>,
    /// Monotonic version; highest signed version wins on conflict.
    pub version: u64,
    /// Designated extension point; unknown keys are preserved, not rejected.
    #[serde(default)]
    pub ext: BTreeMap<String, String>,
}

impl RecordPayload {
    /// The exact canonical bytes that get signed for this payload — the
    /// cross-language ground truth (`domain ‖ 0x00 ‖ CBOR`).
    pub fn signing_input(&self) -> Result<Vec<u8>> {
        signing_bytes(RECORD_DOMAIN, self)
    }

    /// Sign this payload with a working key, producing a [`SignedRecord`].
    pub fn sign(self, working: &WorkingKey) -> Result<SignedRecord> {
        let msg = signing_bytes(RECORD_DOMAIN, &self)?;
        let signature = working.sign(&msg);
        Ok(SignedRecord {
            payload: self,
            signer_pub: working.public(),
            signature,
        })
    }
}

/// A record plus the working-key signature over its payload.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct SignedRecord {
    pub payload: RecordPayload,
    /// The working public key that produced `signature`.
    #[serde(with = "serde_bytes")]
    pub signer_pub: Vec<u8>,
    #[serde(with = "serde_bytes")]
    pub signature: Vec<u8>,
}

impl SignedRecord {
    pub fn id(&self) -> &Id {
        &self.payload.id
    }

    pub fn to_cbor(&self) -> Result<Vec<u8>> {
        let mut buf = Vec::new();
        ciborium::into_writer(self, &mut buf).map_err(|e| Error::Serialization(e.to_string()))?;
        Ok(buf)
    }

    pub fn from_cbor(bytes: &[u8]) -> Result<Self> {
        ciborium::from_reader(bytes).map_err(|e| Error::Serialization(e.to_string()))
    }

    pub fn to_json(&self) -> Result<String> {
        serde_json::to_string(self).map_err(|e| Error::Serialization(e.to_string()))
    }

    pub fn from_json(s: &str) -> Result<Self> {
        serde_json::from_str(s).map_err(|e| Error::Serialization(e.to_string()))
    }

    /// Full verification (plan §7): id binding, root-endorsed and unexpired
    /// working key, not revoked, and a valid signature over the payload.
    pub fn verify(&self, now: u64, revocations: &RevocationSet) -> Result<()> {
        // 1–3. The signer must be a valid working key for this identity.
        verify_working_key(
            &self.payload.root_public_key,
            &self.payload.id,
            &self.payload.keys,
            &self.signer_pub,
            now,
            revocations,
        )?;

        // 4. The signature must verify over the canonical payload bytes.
        let msg = signing_bytes(RECORD_DOMAIN, &self.payload)?;
        verify_raw(&self.signer_pub, &msg, &self.signature)
    }
}
